### PR TITLE
Map should do more work locally before passing up to reduce

### PR DESCRIPTION
**Pull request by corylanou**

### CODE REVIEW

**Comment by corylanou** [comment]
@pauldix

File: influxql/engine.go [hunk 1]
```diff
@@ -669,54 +669,66 @@ func ReduceMean(key Key, values []interface{}, e *Emitter) {
...
-			min = &m
+		v := value.(float64)
+		// Initialize min
+		if min == nil {
+			min = &v
 		}
+		m := math.Min(*min, v)
+		min = &m
 	}
 	e.Emit(key, min)
```
Comment (otoolep) [inline influxql/engine.go]: Do you need to wrap this call to `Emit` in a `if` check for exactly the same reasons as the map phase? If every single node returned zero results, this too would have no results to emit, right?

Comment (corylanou) [inline influxql/engine.go]: That's a good question, and I don't know the answer to that. I have a test that tests that, and it passes, but I haven't looked deep enough to know why it works yet.  Perhaps @pauldix has more insight as to if I'm being to defensive or not enough on this call.

File: influxql/engine.go [hunk 2]
```diff
@@ -669,54 +669,66 @@ func ReduceMean(key Key, values []interface{}, e *Emitter) {
 
 // MapMin collects the values to pass to the reducer
 func MapMin(itr Iterator, e *Emitter, tmin int64) {
-	var values []float64
+	var min *float64
```
Comment (pauldix) [inline influxql/engine.go]: Wouldn't it be easier to just set min to `math.MaxFloat64` and then avoid doing the pointer stuff? Then only emit if `min != math.MaxFloat64`. Of course then you couldn't have that as the min value, but I think that's ok

File: influxql/engine.go [hunk 3]
```diff
@@ -669,54 +669,66 @@ func ReduceMean(key Key, values []interface{}, e *Emitter) {
...
+		m := math.Min(*min, v)
+		min = &m
 	}
 	e.Emit(key, min)
 }
 
 // MapMax collects the values to pass to the reducer
 func MapMax(itr Iterator, e *Emitter, tmax int64) {
-	var values []float64
+	var max *float64
```
Comment (pauldix) [inline influxql/engine.go]: Same comment as before but set max to `math.MaxFloat64 * -1`

File: influxql/engine.go (hunk 1 above)
Comment (pauldix) [inline influxql/engine.go]: He's right, you shouldn't emit a value from the reducer if you never got any values back.

File: influxql/engine.go (hunk 2 above)
Comment (corylanou) [inline influxql/engine.go]: What if the max really is the math.MaxFloat64?

Comment (pauldix) [inline influxql/engine.go]: If you want to ensure we can do that, then have a bool for `pointsYielded` and set that to true. Then only emit if that is true.

Comment (corylanou) [inline influxql/engine.go]: Chicken, meet egg.  Egg, meet chicken. :-)

**Comment by corylanou** [comment]
@pauldix @otoolep removed the pointers, created a flag to check if we yielded data.  take a look.

**Comment by otoolep** [comment]
+1. You're actually back to the boolean approach I suggested at the start, so all good here.

**Comment by corylanou** [comment]
Yeah, most times it simply doesn't pay to be clever.  I should know better by now :-)

**Pull request merged**